File: solutions/javascript/src/structure/dynamic/DynamicLinkedListADT.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class DynamicLinkedListADT {
  constructor() {
    this.head = null;
    this.length = 0;
  }

  // Agrega al final
  add(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
    } else {
      let actual = this.head;
      while (actual.next) {
        actual = actual.next;
      }
      actual.next = newNode;
    }
    this.length++;
  }

  // Inserta en un índice
  insert(index, value) {
    if (index < 0 || index > this.length) {
      throw new Error("Índice fuera de rango");
    }

    const newNode = new Node(value);

    if (index === 0) {
      newNode.next = this.head;
      this.head = newNode;
    } else {
      let actual = this.head;
      for (let i = 0; i < index - 1; i++) {
        actual = actual.next;
      }
      newNode.next = actual.next;
      actual.next = newNode;
    }

    this.length++;
  }

  // Elimina por índice
  remove(index) {
    if (index < 0 || index >= this.length) {
      throw new Error("Índice fuera de rango");
    }

    if (index === 0) {
      this.head = this.head.next;
    } else {
      let actual = this.head;
      for (let i = 0; i < index - 1; i++) {
        actual = actual.next;
      }
      actual.next = actual.next.next;
    }

    this.length--;
  }

  // Obtiene valor por índice
  get(index) {
    if (index < 0 || index >= this.length) {
      throw new Error("Índice fuera de rango");
    }

    let actual = this.head;
    for (let i = 0; i < index; i++) {
      actual = actual.next;
    }
    return actual.value;
  }

  // Retorna el tamaño
  size() {
    return this.length;
  }

  // Comprueba si está vacía
  isEmpty() {
    return this.length === 0;
  }
}

export default DynamicLinkedListADT;

File: solutions/python/structure/dynamic/DynamicLinkedListADT.py
class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.siguiente = None

class DynamicLinkedListADT:
    def __init__(self):
        self.cabeza = None
        self._tamanio = 0

    def add(self, value: int):
        nuevo_nodo = Nodo(value)
        if self.cabeza is None:
            self.cabeza = nuevo_nodo
        else:
            actual = self.cabeza
            while actual.siguiente:
                actual = actual.siguiente
            actual.siguiente = nuevo_nodo
        self._tamanio += 1

    def insert(self, index: int, value: int):
        if index < 0 or index > self._tamanio:
            raise IndexError("Índice fuera de rango")
        nuevo_nodo = Nodo(value)
        if index == 0:
            nuevo_nodo.siguiente = self.cabeza
            self.cabeza = nuevo_nodo
        else:
            actual = self.cabeza
            for _ in range(index - 1):
                actual = actual.siguiente
            nuevo_nodo.siguiente = actual.siguiente
            actual.siguiente = nuevo_nodo
        self._tamanio += 1

    def remove(self, index: int):
        if index < 0 or index >= self._tamanio:
            raise IndexError("Índice fuera de rango")
        if index == 0:
            self.cabeza = self.cabeza.siguiente
        else:
            actual = self.cabeza
            for _ in range(index - 1):
                actual = actual.siguiente
            actual.siguiente = actual.siguiente.siguiente
        self._tamanio -= 1

    def get(self, index: int) -> int:
        if index < 0 or index >= self._tamanio:
            raise IndexError("Índice fuera de rango")
        actual = self.cabeza
        for _ in range(index):
            actual = actual.siguiente
        return actual.valor

    def size(self) -> int:
        return self._tamanio

    def is_empty(self) -> bool:
        return self._tamanio == 0

    def mostrar(self):
        actual = self.cabeza
        while actual:
            print(actual.valor, end=" -> ")
            actual = actual.siguiente
        print("None")
